src/vos/vos_io.c

```diff
@@ -1252,26 +1261,66 @@ vos_update_end(daos_handle_t ioh, uint32_t pm_ver, daos_key_t *dkey, int err)
 	err = vos_publish_blocks(ioc->ic_obj, &ioc->ic_blk_exts, true);
 
 abort:
-	err = err ? umem_tx_abort(umem, err) : umem_tx_commit(umem);
+	if (err == -DER_INPROGRESS && dth != NULL &&
+	    (dth->dth_flags & DTX_F_AOC) &&
+	    !daos_is_zero_dti(&dth->dth_conflict)) {
+		int	rc;
+
+		/* Abort the conflict DTX by force, then try again. */
+		rc = vos_dtx_abort_internal(ioc->ic_obj->obj_cont,
+					    &dth->dth_conflict, 1, true);
+		if (rc == 0) {
+			daos_dti_gen(&dth->dth_conflict, true);
+			goto again;
+		}
+	}
+
+	if (err == 0 && dth != NULL)
+		err = vos_dtx_prepared(dth);
+
+	if (err == 0)
+		umem_tx_commit(umem);
+	else
+		umem_tx_abort(umem, err);
+
 out:
 	if (err != 0)
 		update_cancel(ioc);
 	vos_ioc_destroy(ioc);
-
+	vos_dth_set(NULL);
 	return err;
 }
 
 int
 vos_update_begin(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 		 daos_key_t *dkey, unsigned int iod_nr, daos_iod_t *iods,
-		 daos_handle_t *ioh)
+		 daos_handle_t *ioh, struct daos_tx_handle *dth,
+		 int *dti_cos_count, struct daos_tx_id *dti_cos)
 {
 	struct vos_io_context *ioc;
 	int rc;
 
+	vos_dth_set(dth);
+
+again:
 	rc = vos_ioc_create(coh, oid, false, epoch, iod_nr, iods, false, &ioc);
-	if (rc != 0)
-		return rc;
+	if (rc != 0) {
+		if (rc == -DER_INPROGRESS && *dti_cos_count > 0) {
```

Comment:
How can vos_ioc_create() return a EINPROGRESS? I suppose it only does some initialization work.

Reply:
It may need to create the object.

Reply:
Hmm, we may need reconsider the interface here. since it looks strange to put cos into the update_begin/end?

Reply:
As my understanding, the 'COS' is introduced mainly for obj/key punch, because current obj/key punch is quite different from normal DAOS update, it'll split tree instantly. This approach can make VOS aggregation faster, but it may complicate DAOS protocol. I think here the 'COS' for 2PC is an example, if we switch to 'log mode' (maintain obj/key punch records just like record punch) obj/key punch, I suppose this 'COS' could be totally removed then, not sure if we are on same page on this?

I think we'd have a discussion over this (switch to log mode obj/key punch?) in the upcoming f2f meeting.

Reply:
On some degree, I would yes. Because of current punch implementation, the other medication cannot share object/key path with punch, as to we have to commit the punch operation before other modification if detect conflict.